File: Stream of Characters.py
# Reverse the words so that instead of searching for suffix we need to search for prefix which is comparatively easy.
class TrieNode:
    def __init__(self):
        self.children = {}
        self.endOfWord = False

class Trie:
    def __init__(self):
        self.root = self.getTrieNode()
    
    def getTrieNode(self):
        return TrieNode()
    
    def insert(self, word):
        root = self.root
        for ch in word:
            if ch not in root.children:
                root.children[ch] = self.getTrieNode()
            root = root.children[ch]
        root.endOfWord = True
    
    def isPrefix(self, word):
        root = self.root
        for i in range(len(word)-1,-1,-1):
            if root.endOfWord:
                return True
            if word[i] not in root.children:
                return False
            root = root.children[word[i]]
        return root.endOfWord
        
class StreamChecker:
    def __init__(self, words: List[str]): #O(Length of all the words)
        self.stream = []
        self.trie = Trie()
        for word in words:
            
            self.trie.insert(word[::-1])
 
    def query(self, letter: str) -> bool: #O(1)
        self.stream.append(letter)
        return self.trie.isPrefix(self.stream)

File: StreamOfCharacters.java
public class StreamOfCharacters {
    class TrieNode{
        boolean isEnd;
        TrieNode[] children;
        private TrieNode(){
            this.children = new TrieNode[26];
        }
    }

    private void insert(String word){
        TrieNode curr = root;
        for(int i=word.length()-1;i>=0;i--){
            char c = word.charAt(i);
            if(curr.children[c-'a']==null){
                curr.children[c-'a'] = new TrieNode();
            }
            curr = curr.children[c-'a'];
        }
        curr.isEnd = true;
    }

    TrieNode root;
    StringBuilder sb;
    int max; // length of the longest word from list of words
    public StreamChecker(String[] words) {
        this.root = new TrieNode();
        this.sb = new StringBuilder();
        for(String word:words){
            insert(word);
            max = Math.max(max, word.length());
        }
    }
    
    public boolean query(char letter) {
        sb.append(letter);
        if(sb.length()>max){
            sb.deleteCharAt(0);
        }
        TrieNode curr = root;
        for(int i=sb.length()-1;i>=0;i--){
            char c = sb.charAt(i);
            if(curr.children[c-'a']==null) return false;
            curr = curr.children[c-'a'];
            if(curr.isEnd) return true;
        }
        return false;
    }
}

// TC - O(n)
// SC - O(n) n - length of longest word